all/HOPE-3--Dependency-Injection.md

```diff
@@ -58,18 +58,18 @@ def my_endpoint(timer: hug.directives.timer):
 
 ### Defining a dependency
 
-Defining a dependency will happen via a new hug `provides` decorator:
+Defining a dependency will happen via a new hug `provide` decorator:
 
 ```python
-    @hug.provides()
+    @hug.provide
```

Comment:
Currently hug's decorators are not currently optionally callable. To support that requires additional changes to HUG's code. For example:
https://chase-seibert.github.io/blog/2013/12/17/python-decorator-optional-parameter.html#optional-parameters
https://realpython.com/primer-on-python-decorators/#both-please-but-never-mind-the-bread

I think it might be neat to explore potentially adding support to HUG, but that should be a discussion for a separate HOPE, not edited in passing in this one.

Reply:
I'm borrowing the optional callability from the pytest fixture pattern.

I'd consider this an implementation detail, which a HOPE probably shouldn't be concerned with.

Reply:
You can see how PyTest had to check for the `callable` nature of the first parameter here:
https://github.com/pytest-dev/pytest/blob/master/src/_pytest/fixtures.py#L1036-L1041

Part of the reason why I'm highlighting this here is that you also converted `@hug.http()` below to `@hug.http` which is out of scope for this HOPE I believe. Converting a decorator to be optionally callable adds some constraints on the nature of the parameters passed into the decorator. If those constraints are acceptable I think it would make sense to apply the pattern more broadly throughout HUG. I just want to make sure the tradeoffs are considered before changing the behavior.

Reply:
> Converting a decorator to be optionally callable adds some constraints on the nature of the parameters passed into the decorator. If those constraints are acceptable I think it would make sense to apply the pattern more broadly throughout HUG. I just want to make sure the tradeoffs are considered before changing the behavior.

Not sure what you mean by "constraints". The pytest fixture decorator, for example, is optionally callable, and provides a fairly full API of passable kwargs as described in the [API doc here](http://doc.pytest.org/en/latest/reference.html#pytest-fixture). We wouldn't necessarily have to implement it the same way as pytest.

Ideally this HOPE would be a specification of the exact API we'd like to implement.

I'm all for exploring implications of our APIs, and I'm not sure what Tim's timeline is, but I think there's plenty of time to explore ramifications of these HOPE docs. Is there a specific experiment you want to run, or a specific question you want to ask?

Reply:
By constraints I mean you have to leave yourself some way to distinguish a decorator call that is receiving the wrapped function and a decorator that is being called directly by the programmer (that needs to return a function that will wrap the the called function). The links above show some of the ways that you can tell between those use cases. Pytest for example makes sure that the first parameter can only be a string when it is being called by a programmer so that that it can distinguish when the first parameter is instead a `callable`, aka the function to wrap.

```python
def decorator(function_or_string="default", optional_param1=None, optional_param2=42):
   pass
```

```python
@decorator
def function():
```
Here `decorator` is being called with `function` being passed into the first parameter (`function_or_string`). Pytest checks for this by seeing if `callable(function_or_string)` as well as verifying that some of the optional parameters are their default values. In this case you need to return the wrapped function directly.

```python
@decorator()
def function():
```
Here instead `decorator` is being passed no parameters so `function_or_string` instead receives `default`. In this case you need to return a function to return the wrapped function.

The two different ways of executing `decorator` result in different behaviors that you have to have additional logic to detect and behave properly. It is definitely doable, but to make sure you can detect which way you are being executed as a decorator means you have to constrain your parameter types. Specifically you can't accept a function (or class with `def __call__()`) as the first parameter because then you can't distinguish the way in which you are being called.

Reply:
Okay, that makes sense. To summarize, we'd have trouble with any kind of decorator that accepted a single positional arg, with that arg being a callable.

e.g. This probably couldn't be implemented

```
@my_optionally_callable_decorator(some_random_callable) 
def wrapped():
    pass
```

Reply:
Yup, and there's additional work if you accept multiple parameters and want to make sure they are all in their default state. By no means impossible, but something that we want to go into with eyes open if we're going to make all of hug's decorators optionally callable (and the more decorators support that workflow the more confusion if some decorators don't support it).